Framework for PEFT via mixins

arendu opened the pull request:
# What does this PR do ?

PTuning and Adapters/IA3 have been implemented in diverging styles. Adapters and IA3 uses the elegant adapter_mixins framework. This PR implements P-Tuning via the same adapter_mixins framework.

The high level goal is to unify all PEFT methods to use adapter_mixins. This is allow us to the GPTSFTDataset to be used for all future PEFT methods.

Currently this PR does not remove the GPTPromptLearingModel functionality, but once adaper-mixin based p-tuning has been tested we can deprecate GPTPromptLearingModel GPTPromptLearningDataset classes.

PTuning does not need to flexibility to insert prompt tokens anywhere in the input. This feature has been removed for simplicity.

Also added:
- mix-n-match PEFT scheme like `adapter_and_ptuning` can be easily extended to do `ia3_and_ptuning` or `lora_and_ptuning`
- grad clipping functionality for PEFT training
- LoRA adapters added for GPT


**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

titu1994 commented on nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py:
```diff
@@ -347,3 +348,43 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
+        self.adapter_name_keys = [AdapterName.PTUNING_ADAPTER]
+        for _, layer in self.frozen_model.named_modules():
+            if hasattr(layer, 'activations_checkpoint_method'):
+                layer.activations_checkpoint_method = (
```
Maybe log this modification ?

titu1994 commented on nemo/collections/nlp/models/language_modeling/megatron_gpt_adapter_model.py:
```diff
@@ -347,3 +348,43 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
+
+        logging.info(f'Before adding adapters:\n{self.frozen_model.summarize()}')
+
+        adapter_cfg = PromptEncoderAdapterConfig(
```
Maybe easier to do PromptEncoderAdapterConfig(**cfg.prompt_encoder_adapter)?

titu1994 commented on nemo/collections/nlp/modules/common/megatron/adapters/parallel_adapters.py:
```diff
@@ -169,3 +172,88 @@ class ParallelLinearAdapterConfig:
+        embedding_dim: int,
+        init_std: float,
+        output_dim: int,
+        adapter_strategy: adapter_mixin_strategies.ResidualAddAdapterStrategyConfig = None,
```
Prompting doesn't use residual right? Or is this totally ignored ?

titu1994 reviewed (commented):
This approach looks great wrt adapter standardization. I'll leave final review to @MaximumEntropy

titu1994 reviewed (commented):
Overall it looks great. The connector is a bit under documented, cause there's a lot of cases it is covering

arendu commented on nemo/collections/nlp/models/language_modeling/megatron_gpt_peft_models.py:
```diff
@@ -0,0 +1,401 @@
+            # at this stage only adapter params will appear in the state_dict arg, so we only update those while the rest of the model is frozen.
+            # setting strict=False will ignore the missing keys (which are not being updated anyway)
+            assert set(state_dict.keys()) == self.adapter_keys
+            super().load_state_dict(state_dict, strict=False)
```
Thats why I check the assert just before load_state_dict.

arendu commented on nemo/collections/nlp/models/language_modeling/megatron_gpt_peft_models.py:
```diff
@@ -0,0 +1,401 @@
+        """ 
+        Gets the keys associated with the adapters only.
+        """
+        state_dict = self.model.state_dict(prefix="model.")
```
Is it? Its in the docs for all named buffers like named_parameters and named_modules etc...

arendu commented:
> Overall it looks great. The connector is a bit under documented, cause there's a lot of cases it is covering

@titu1994  Many thanks for looking into this PR! I Added docs to PEFTSaveRestoverConnector and addressed other issues you raised.

titu1994 reviewed (approved):
Overall looks good. Requires final review from @ericharper for the new save restore connector part but to me that looks reasonable.

yidong72 commented on nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:
```diff
@@ -41,9 +41,11 @@ def __init__(
         index_mapping_dir: str = None,
         prompt_template: str = None,
+        virtual_tokens: int = 0,
+        tokens_to_generate: int = 0,
```
add documents for `virtual_tokens` and `tokens_to_generate`

yidong72 commented on nemo/collections/nlp/data/language_modeling/megatron/gpt_sft_dataset.py:
```diff
@@ -41,9 +41,11 @@ def __init__(
         answer_only_loss: bool = True,
         truncation_field: str = "answer",
-        pad_to_max_length: bool = True,
+        pad_to_max_length: bool = False,  # (@adithyare) allows for much faster training especially in PEFT settings.
```
don't change the default for backward compatibility just in case others rely on the default behavior

arendu replied:
fixing..

yidong72 commented on nemo/collections/nlp/models/language_modeling/megatron_gpt_peft_models.py:
```diff
@@ -0,0 +1,361 @@
+        """
+        ModelPT override. Optimizer will get self._optimizer_param_groups. 
+        Makes two optimizer param groups, one for the frozen model params
+        and one for the prompt-table/prompt-encoder params. The learning 
```
the comment is outdated. need to update.

yidong72 commented on nemo/collections/nlp/models/language_modeling/megatron_gpt_sft_model.py:
```diff
@@ -83,6 +89,7 @@ def __init__(self, cfg: DictConfig, trainer: Trainer):
         self.original_checkpointing_granularity = base_module.language_model.encoder.activations_checkpoint_granularity
         self.original_checkpointing_num_layers = base_module.language_model.encoder.activations_checkpoint_num_layers
         self.original_checkpointing_method = base_module.language_model.encoder.activations_checkpoint_method
+        self.virtual_tokens = 0
```
do you have to have the virtual tokens here? why not read it from the cfg like the `tokens_to_generate`?

yidong72 commented on examples/nlp/language_modeling/tuning/megatron_gpt_peft_eval.py:
```diff
@@ -0,0 +1,157 @@
+    if model.global_rank == 0:
+        print("***************************")
+        if cfg.inference.outfile_path is not None:
+            with open(cfg.inference.outfile_path, "w", encoding="utf-8") as f:
```
add the inference server like gpt_eval script?

arendu replied:
Definitely want to support this. I'll add it in a separate PR.

yidong72 reviewed (commented):
Great to see everything is merged under the adaptor framework. Just have some minor comments.
